sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -364,7 +364,12 @@ case class DataSource(
             catalogTable.get,
             catalogTable.get.stats.map(_.sizeInBytes.toLong).getOrElse(defaultTableSize))
         } else {
-          new InMemoryFileIndex(sparkSession, globbedPaths, options, Some(partitionSchema))
```

Comment:
I'd like to create file status cache as a local variable, pass it to `getOrInferFileFormatSchema`, then use it here. It's much easier to reason about the lifetime of this cache by this way.